### PR TITLE
Create a lexicographic comparator operator and replace uses in capnproto

### PR DESCRIPTION
**Pull request by AdityaAtulTewari**

### CODE REVIEW

File: c++/src/kj/common.h [hunk 1]
```diff
@@ -1928,6 +1928,20 @@ class ArrayPtr: public DisallowConstCopyIfNotConst<T> {
     return true;
   }
 
+  //Lexographic difference function can be overriden for the type by defining <=> on your type
```
Comment (mikea) [inline c++/src/kj/common.h:1931]: nits: //<space>

also check around for comment style - it comes _after_ the line. (kj style guide)

File: c++/src/kj/common.h [hunk 2]
```diff
@@ -1928,6 +1928,20 @@ class ArrayPtr: public DisallowConstCopyIfNotConst<T> {
...
+  //Designed infer the size to compare as the minimum i.e. auto truncates the comparison
+  inline int operator<=>(const ArrayPtr& other) const {
+    size_t comparisonSize = kj::min(size_, other.size_);
```
Comment (mikea) [inline c++/src/kj/common.h:1934]: well, this is wrong. it thingks "a"  euqals "aa".

File: c++/src/kj/common.h [hunk 3]
```diff
@@ -1928,6 +1928,20 @@ class ArrayPtr: public DisallowConstCopyIfNotConst<T> {
     return true;
   }
 
+  //Lexographic difference function can be overriden for the type by defining <=> on your type
+  //Designed infer the size to compare as the minimum i.e. auto truncates the comparison
+  inline int operator<=>(const ArrayPtr& other) const {
```
Comment (mikea) [inline c++/src/kj/common.h:1933]: tests please

File: c++/src/kj/common.h [hunk 4]
```diff
@@ -1922,12 +1923,57 @@ class ArrayPtr: public DisallowConstCopyIfNotConst<T> {
   template <typename U>
   inline bool operator==(const ArrayPtr<U>& other) const {
     if (size_ != other.size()) return false;
-    for (size_t i = 0; i < size_; i++) {
+    for (size_t i = 1; i < size_; i++) {
```
Comment (AdityaAtulTewari) [inline c++/src/kj/common.h:1926]: This was a mistake.

File: c++/src/kj/common.h (hunk 3 above)
Comment (AdityaAtulTewari) [reply c++/src/kj/common.h]: Please let me know if changes to: c++/src/kj/common-test.c++ are sufficient for this now.

**Pull request closed without merging**